### PR TITLE
Modified ScoreServicesImpl to also show outputs in  CLI

### PR DESCRIPTION
**Pull request by tudorlesan**
Signed-off-by: Lesan Tudor <tudor-andrei.lesan@hp.com>

### CODE REVIEW

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java [hunk 1]
```diff
@@ -131,6 +131,37 @@ public synchronized void onEvent(ScoreEvent scoreEvent) throws InterruptedExcept
                         printWithColor(Ansi.Color.YELLOW, prefix + taskName);
                     }
                     break;
+                case ScoreLangConstants.EVENT_OUTPUT_END:
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: all the logic must be in a diff method..
moreover.. i do not understand what you are doing here, why so much parsing is needed?

**Comment by meirwah** [comment]
can you add an image snapshot on how it looks to run this in the CLI?

**Comment by TudorLesanHP** [comment]
cli_outputs

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java (hunk 1 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: I first make sure I print just the actual outputs not also the published ones. Then get the prefix (nr. of '-'s)
When printing the outputs I truncate the value if the length is greater than 30, or put "(empty)" if output has no value.
I also replace newlines and then print the actual output (key - value)

**Comment by TudorLesanHP** [comment]
fixes #5

**Comment by orius123** [comment]
I think it's better to present only the outputs of the main flow and not all of the children.

**Comment by meirwah** [comment]
agree with Eliya

**Comment by TudorLesanHP** [comment]
@orius123 @meirwah , only main flow outputs are shown now

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java [hunk 2]
```diff
@@ -124,19 +124,51 @@ public synchronized void onEvent(ScoreEvent scoreEvent) throws InterruptedExcept
...
         }
 
+        private void printOutputs(Map<String, Serializable> data) {
+            if(data.containsKey(LanguageEventData.OUTPUTS) && data.containsKey(LanguageEventData.PATH) && data.get(LanguageEventData.PATH).equals("0/0")) {
+
+                @SuppressWarnings("unchecked") Map<String, String> outputs = (Map<String, String>) data.get(LanguageEventData.OUTPUTS);
+
+                String prefix = getPrefix(data);
+
+                if (outputs != null && !outputs.keySet().isEmpty()) {
```
Comment (orius123) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: Would use apache map utils

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java [hunk 3]
```diff
@@ -124,19 +124,51 @@ public synchronized void onEvent(ScoreEvent scoreEvent) throws InterruptedExcept
...
+                    printWithColor(Ansi.Color.WHITE, prefix + "Outputs:");
+
+                    for (String key : outputs.keySet()) {
+                        String outputValue = outputs.get(key).replace("\n", " ");
+                        outputs.put(key, StringUtils.abbreviate(outputValue, 0, OUTPUT_VALUE_LIMIT));
+                        if (StringUtils.isEmpty(outputValue)) {
+                            outputs.put(key, "(empty)");
+                        }
+                        printWithColor(Ansi.Color.WHITE, prefix + "  " + key + " = " + outputs.get(key));
+                    }
```
Comment (orius123) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: Do we still need the prefix if we only print the flow outputs?

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java [hunk 4]
```diff
@@ -124,19 +124,51 @@ public synchronized void onEvent(ScoreEvent scoreEvent) throws InterruptedExcept
...
+
                 case EVENT_EXECUTION_FINISHED :
                     printFinishEvent(data);
                     break;
             }
         }
 
+        private void printOutputs(Map<String, Serializable> data) {
+            if(data.containsKey(LanguageEventData.OUTPUTS) && data.containsKey(LanguageEventData.PATH) && data.get(LanguageEventData.PATH).equals("0/0")) {
+
```
Comment (orius123) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: Please use a constant for the root path

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java (hunk 3 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: no we don't need it anymore. Though I suggest printing like we were to get the prefix so it looks better.

**Comment by meirwah** [comment]
i think we can skip the "Outputs" keyword print, enough is to just print the outputs, WDYT?

**Comment by TudorLesanHP** [comment]
fine by me.

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java [hunk 5]
```diff
@@ -124,19 +128,48 @@ public synchronized void onEvent(ScoreEvent scoreEvent) throws InterruptedExcept
...
+        private void printOutputs(Map<String, Serializable> data) {
+            if(data.containsKey(LanguageEventData.OUTPUTS) && data.containsKey(LanguageEventData.PATH) && data.get(LanguageEventData.PATH).equals(EXEC_START_PATH)) {
+
+                @SuppressWarnings("unchecked") Map<String, String> outputs = (Map<String, String>) data.get(LanguageEventData.OUTPUTS);
+                Map<String, String> modifiableOutputs = new HashMap<>();
+                modifiableOutputs.putAll(outputs);
+                if (MapUtils.isNotEmpty(outputs)) {
+
+                    for (String key : outputs.keySet()) {
+                        String outputValue = modifiableOutputs.get(key).replace("\n", " ");
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: this .replace("\n", " "), is error prone.. what if the value of the output is "n\n" ?
also where are the tests for this?

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java [hunk 6]
```diff
@@ -124,19 +128,48 @@ public synchronized void onEvent(ScoreEvent scoreEvent) throws InterruptedExcept
...
+                        modifiableOutputs.put(key, StringUtils.abbreviate(outputValue, 0, OUTPUT_VALUE_LIMIT));
+                        if (StringUtils.isEmpty(outputValue)) {
+                            modifiableOutputs.put(key, "(empty)");
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: i think if the output is empty, just skip it...

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java [hunk 7]
```diff
@@ -124,19 +128,48 @@ public synchronized void onEvent(ScoreEvent scoreEvent) throws InterruptedExcept
...
                     printFinishEvent(data);
                     break;
             }
         }
 
+        private void printOutputs(Map<String, Serializable> data) {
+            if(data.containsKey(LanguageEventData.OUTPUTS) && data.containsKey(LanguageEventData.PATH) && data.get(LanguageEventData.PATH).equals(EXEC_START_PATH)) {
+
+                @SuppressWarnings("unchecked") Map<String, String> outputs = (Map<String, String>) data.get(LanguageEventData.OUTPUTS);
+                Map<String, String> modifiableOutputs = new HashMap<>();
```
Comment (meirwah) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: all this (lines 150-152) can be shorten to a simple : 
Map<String, String> outputs = new HashMap((Map<String, String>)data.get(LanguageEventData.OUTPUTS))

File: score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java (hunk 5 above)
Comment (TudorLesanHP) [inline score-lang-cli/src/main/java/org/openscore/lang/cli/services/ScoreServicesImpl.java]: I can remove this then, but outputs may be oddly shown in the CLI then.

**Comment by orius123** [comment]
@tudorlesan made some cosmetic fixes and merged from master.
what happens in cases where outputs are not strings?

**Comment by tudorlesan** [comment]
I tested with this solution and if a flow output was not String it threw an error. So I changed the Map from <String,String> to <String,Serializable> and everything seems to be working fine. When checking empty values or too long values the toString() is applied.

**Comment by oritstone** [comment]
@tudorlesan is this pull request ready for review & merge?

**Comment by tudorlesan** [comment]
@oritstone yes, it's ready

**Pull request merged**